File: Sources/Olaf/UI/Model/LogViewerModel.swift
import Foundation
import Combine

/// State owner for the viewer: merges the instantaneous snapshot with the live stream, and applies
/// level/category/text filtering.
@MainActor
public final class LogViewerModel: ObservableObject {

    /// All collected entries (snapshot + live).
    @Published public private(set) var entries: [LogEntry] = []

    /// Display scope.
    public enum Scope: String, CaseIterable, Sendable {
        case session   // this session (in-memory ring buffer)
        case history   // entire disk history (including previous sessions)
    }
    @Published public var scope: Scope = .session

    /// Filters.
    @Published public var searchText: String = ""
    /// Debounced, normalized (trim + lowercase) form of `searchText`. Updated not on every
    /// keystroke, but after the user stops typing → we don't scan the whole list on every
    /// keystroke. `filteredEntries` uses this.
    @Published private var effectiveQuery: String = ""
    /// Levels to display (multi-select). Empty = all.
    @Published public var enabledLevels: Set<LogLevel> = Set(LogLevel.allCases)
    @Published public var selectedCategories: Set<LogCategory> = []
    /// Network content-type filter (empty = off). When selected, **only** network entries of
    /// these types are shown (non-network entries are hidden).
    @Published public var selectedContentKinds: Set<NetworkContentKind> = []

    /// `true` while new logs are appended live; `false` (paused) freezes the list.
    @Published public var isFollowing: Bool = true

    /// Pinned entry identifiers (per-session; not persisted). Pins are shown in a separate
    /// section at the top of the list, **independent of filters**.
    @Published public var pinnedIDs: Set<UUID> = []

    private var pendingWhilePaused: [LogEntry] = []
    private var streamTask: Task<Void, Never>?
    private var cancellables = Set<AnyCancellable>()

    /// Live-stream coalescing buffer: incoming entries aren't appended to `entries` one at a
    /// time (each append triggers an `@Published` emission → a full SwiftUI diff + recomputing
    /// `filteredEntries`); instead they accumulate here and are flushed **in bulk** at a short
    /// interval. This significantly reduces main-thread churn under heavy logging (network capture).
    private var incoming: [LogEntry] = []
    private var flushScheduled = false
    private static let coalesceInterval: UInt64 = 100_000_000  // 100 ms

    /// `true` while history (disk) is loading.
    @Published public private(set) var isLoading: Bool = false

    // MARK: - History pagination
    //
    // History is now loaded page by page (newest to oldest) rather than all at once. Filter/search
    // only operates on LOADED entries; as the user scrolls (or taps the button), older pages are
    // appended to the end of the list.

    /// Is there an older, not-yet-loaded page on disk?
    @Published public private(set) var hasMoreHistory: Bool = false
    /// `true` while an older page is loading (the row below the list shows a spinner).
    @Published public private(set) var isLoadingMore: Bool = false
    /// Cursor for the next (older) page.
    private var historyCursor: String?
    /// Target minimum entry count per page.
    private static let historyPageSize = 500

    // MARK: - Derived (memoized) state
    //
    // These used to be computed properties → the entire list was rescanned on every SwiftUI
    // render (stutter with large History). Now they are computed once, only when the inputs
    // change, and published via @Published.

    /// Filtered entries, sorted newest first.
    @Published public private(set) var filteredEntries: [LogEntry] = []

    /// Decode errors folded under their network entry (list shows a badge on the
    /// network row instead of one row per decode error; detail lists them all).
    @Published private(set) var decodeIndex: DecodeAttachmentIndex = .empty

    /// Session groups for history entries — **excluding the current session** (shown in the "Session" tab).
    @Published public private(set) var sessionGroups: [LogSession] = []

    /// Categories seen in the current entries (for the filter bar).
    @Published public private(set) var availableCategories: [LogCategory] = []

    public init() {
        // Search debounce: searchText → (trim+lowercase) → 200ms of silence → effectiveQuery.
        $searchText
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() }
            .removeDuplicates()
            .debounce(for: .milliseconds(200), scheduler: DispatchQueue.main)
            .assign(to: &$effectiveQuery)

        // Values derived from inputs: computed only when an input changes (not per render).
        // The decode index is built in the same pass as the filter so the two published
        // values can never disagree (a row hidden by one but not badged by the other).
        let categorySelection = Publishers.CombineLatest($selectedCategories, $selectedContentKinds)
        let derived = Publishers.CombineLatest4($entries, $effectiveQuery, $enabledLevels, categorySelection)
            .map { entries, query, levels, selection -> (filtered: [LogEntry], index: DecodeAttachmentIndex) in
                let index = DecodeAttachmentIndex.build(from: entries)
                let filtered = Self.filter(
                    entries: entries, query: query, levels: levels,
                    categories: selection.0, contentKinds: selection.1,
                    decodeIndex: index
                )
                return (filtered, index)
            }
            .share()
        derived.map(\.filtered).assign(to: &$filteredEntries)
        derived.map(\.index).assign(to: &$decodeIndex)

        $filteredEntries
            .map { Self.groupSessions($0, excluding: Olaf.currentSessionID) }
            .assign(to: &$sessionGroups)

        $entries
            .map { Self.categories(in: $0) }
            .removeDuplicates()
            .assign(to: &$availableCategories)

        // Default filter: the viewer opens with the network category preselected — but only
        // when network entries actually exist, because the chip bar only renders seen
        // categories: preselecting an unseen one would blank the list behind an invisible,
        // chipless filter. Applied once, on the first non-empty load, and never over a
        // selection the user already made.
        $entries
            .filter { !$0.isEmpty }
            .first()
            .sink { [weak self] entries in
                guard let self, self.selectedCategories.isEmpty else { return }
                self.selectedCategories = Self.defaultCategorySelection(for: entries)
            }
            .store(in: &cancellables)
    }

    deinit { streamTask?.cancel() }

    // MARK: - Lifecycle

    public func start() {
        reload()
        guard streamTask == nil else { return }
        streamTask = Task { [weak self] in
            for await entry in Olaf.stream() {
                guard let self else { return }
                self.append(entry)
            }
        }
    }

    /// (Re)loads entries according to scope. In both session (memory) and history (disk) mode,
    /// reading is done **asynchronously** → the UI/main thread never blocks (even under heavy
    /// core-queue load). In history mode only the FIRST page is loaded; the rest comes via `loadOlderHistory()`.
    public func reload() {
        pendingWhilePaused.removeAll()
        incoming.removeAll()
        historyCursor = nil
        hasMoreHistory = false
        isLoadingMore = false
        switch scope {
        case .session:
            Task { [weak self] in
                let snapshot = await Olaf.snapshotAsync()
                guard let self else { return }
                self.entries = snapshot
            }
        case .history:
            isLoading = true
            Task { [weak self] in
                let page = await Olaf.loadPersistedPage(minimumEntries: Self.historyPageSize)
                guard let self else { return }
                self.entries = page.entries
                self.historyCursor = page.nextCursor
                self.hasMoreHistory = page.nextCursor != nil
                self.isLoading = false
            }
        }
    }

    /// Loads the next (older) page of history and appends it to the end of the list.
    /// Idempotent: calls are a no-op while loading, or when there is no cursor.
    public func loadOlderHistory() {
        guard scope == .history, !isLoading, !isLoadingMore, let cursor = historyCursor else { return }
        isLoadingMore = true
        Task { [weak self] in
            let page = await Olaf.loadPersistedPage(before: cursor, minimumEntries: Self.historyPageSize)
            guard let self else { return }
            // Don't apply the result if the page was reset by a reload in the meantime (scope change, etc).
            guard self.scope == .history, self.historyCursor == cursor else {
                self.isLoadingMore = false
                return
            }
            // `entries` is kept oldest-to-newest; the older page is inserted at the FRONT.
            self.entries.insert(contentsOf: page.entries, at: 0)
            self.historyCursor = page.nextCursor
            self.hasMoreHistory = page.nextCursor != nil
            self.isLoadingMore = false
        }
    }

    public func setScope(_ newScope: Scope) {
        guard newScope != scope else { return }
        scope = newScope
        reload()
    }

    private func append(_ entry: LogEntry) {
        incoming.append(entry)
        scheduleFlush()
    }

    /// Flushes the accumulated `incoming` entries **all at once** after a short delay (coalescing).
    private func scheduleFlush() {
        guard !flushScheduled else { return }
        flushScheduled = true
        Task { [weak self] in
            try? await Task.sleep(nanoseconds: Self.coalesceInterval)
            guard let self else { return }
            self.flushScheduled = false
            guard !self.incoming.isEmpty else { return }
            // Live entries are not mixed into the list while in history scope (the current
            // session isn't shown there anyway); reload() picks up the latest snapshot when
            // returning to Session — nothing is lost.
            if self.scope == .history {
                self.incoming.removeAll(keepingCapacity: true)
                return
            }
            // If paused, entries wait until "resume" (pause semantics preserved).
            if self.isFollowing {
                self.entries.append(contentsOf: self.incoming)
            } else {
                self.pendingWhilePaused.append(contentsOf: self.incoming)
            }
            self.incoming.removeAll(keepingCapacity: true)
        }
    }

    public func resumeFollowing() {
        isFollowing = true
        if !pendingWhilePaused.isEmpty {
            entries.append(contentsOf: pendingWhilePaused)
            pendingWhilePaused.removeAll()
        }
    }

    // MARK: - Derivation logic (pure functions → testable)

    /// Grouped logs for one app session.
    public struct LogSession: Identifiable, Sendable {
        public let id: String        // sessionID
        public let startDate: Date   // earliest entry in the session
        public let entries: [LogEntry]  // newest first
    }

    /// Level/category/content-type/search filter — newest first.
    ///
    /// Decode entries attached to a network entry are never listed as rows; their
    /// network entry answers for them instead: it matches the `.decoding` category
    /// chip, the `.error` level filter, and search queries that hit an attached
    /// entry — so folding never makes a decode failure unfindable.
    nonisolated static func filter(
        entries: [LogEntry],
        query: String,   // already trim+lowercase (from the debounce pipeline)
        levels: Set<LogLevel>,
        categories: Set<LogCategory>,
        contentKinds: Set<NetworkContentKind> = [],
        decodeIndex: DecodeAttachmentIndex = .empty
    ) -> [LogEntry] {
        entries.reversed().filter { entry in
            if decodeIndex.attachedIDs.contains(entry.id) { return false }
            let attachedErrors = decodeIndex.errors(for: entry)
            if !levels.isEmpty, !levels.contains(entry.level) {
                guard !attachedErrors.isEmpty, levels.contains(.error) else { return false }
            }
            if !categories.isEmpty, !categories.contains(entry.category) {
                guard !attachedErrors.isEmpty, categories.contains(.decoding) else { return false }
            }
            if !contentKinds.isEmpty {
                guard let kind = NetworkContentKind.of(entry), contentKinds.contains(kind) else {
                    return false
                }
            }
            guard !query.isEmpty else { return true }
            return matchesQuery(entry, query) || attachedErrors.contains { matchesQuery($0, query) }
        }
    }

    nonisolated private static func matchesQuery(_ entry: LogEntry, _ query: String) -> Bool {
        if entry.message.lowercased().contains(query) { return true }
        if entry.category.rawValue.lowercased().contains(query) { return true }
        return entry.metadata.contains { key, value in
            key.lowercased().contains(query) || value.lowercased().contains(query)
        }
    }

    /// Groups entries by session; `excluding` (the current session) is skipped.
    /// Sessions are sorted newest to oldest; their contents are newest first.
    nonisolated static func groupSessions(_ entries: [LogEntry], excluding current: String) -> [LogSession] {
        var grouped: [String: [LogEntry]] = [:]
        var order: [String] = []
        for entry in entries where entry.sessionID != current {
            if grouped[entry.sessionID] == nil { order.append(entry.sessionID) }
            grouped[entry.sessionID, default: []].append(entry)
        }
        return order
            .map { id in
                let items = grouped[id] ?? []
                let start = items.map(\.date).min() ?? .distantPast
                return LogSession(id: id, startDate: start, entries: items)
            }
            .sorted { $0.startDate > $1.startDate }
    }

    /// Initial category selection for a fresh viewer: `.network` preselected when network
    /// entries exist, nothing otherwise (all categories shown).
    nonisolated static func defaultCategorySelection(for entries: [LogEntry]) -> Set<LogCategory> {
        entries.contains { $0.category == .network } ? [.network] : []
    }

    /// Categories seen in the entries (sorted by name, deduplicated).
    nonisolated static func categories(in entries: [LogEntry]) -> [LogCategory] {
        var seen = Set<LogCategory>()
        var ordered: [LogCategory] = []
        for entry in entries where !seen.contains(entry.category) {
            seen.insert(entry.category)
            ordered.append(entry.category)
        }
        return ordered.sorted { $0.rawValue < $1.rawValue }
    }

    // MARK: - Actions

    public func toggleCategory(_ category: LogCategory) {
        if selectedCategories.contains(category) {
            selectedCategories.remove(category)
        } else {
            selectedCategories.insert(category)
        }
    }

    public func toggleLevel(_ level: LogLevel) {
        if enabledLevels.contains(level) {
            enabledLevels.remove(level)
        } else {
            enabledLevels.insert(level)
        }
    }

    public func togglePin(_ entry: LogEntry) {
        if pinnedIDs.contains(entry.id) {
            pinnedIDs.remove(entry.id)
        } else {
            pinnedIDs.insert(entry.id)
        }
    }

    /// Pinned entries (newest first) — independent of filters, from among the loaded entries.
    public var pinnedEntries: [LogEntry] {
        Self.pinned(in: entries, ids: pinnedIDs)
    }

    nonisolated static func pinned(in entries: [LogEntry], ids: Set<UUID>) -> [LogEntry] {
        guard !ids.isEmpty else { return [] }
        return entries.reversed().filter { ids.contains($0.id) }
    }

    /// Writes the selected entries (multi-select) to a readable `.log` file — in chronological order.
    public func exportFileURL(for ids: Set<UUID>) async -> URL? {
        let chosen = Array(filteredEntries.filter { ids.contains($0.id) }.reversed())
        guard !chosen.isEmpty else { return nil }
        return await Olaf.exportFileURL(entries: chosen)
    }

    public func toggleContentKind(_ kind: NetworkContentKind) {
        if selectedContentKinds.contains(kind) {
            selectedContentKinds.remove(kind)
        } else {
            selectedContentKinds.insert(kind)
        }
    }

    /// Is a non-default filter applied? (shown in the funnel badge)
    public var isFiltering: Bool {
        enabledLevels.count != LogLevel.allCases.count
            || !selectedCategories.isEmpty
            || !selectedContentKinds.isEmpty
    }

    public func resetFilters() {
        enabledLevels = Set(LogLevel.allCases)
        selectedCategories.removeAll()
        selectedContentKinds.removeAll()
    }

    public func clear() {
        Olaf.clear()
        entries.removeAll()
        pendingWhilePaused.removeAll()
        incoming.removeAll()
        pinnedIDs.removeAll()
    }

    /// Writes the entries **currently visible on screen** (scope + level/category/search filters
    /// applied) to a shareable file. If no filter is selected, this equals the entire visible
    /// list. Written chronologically (oldest → newest); `filteredEntries` is sorted newest first.
    public func exportFileURL() async -> URL? {
        let visible = Array(filteredEntries.reversed())
        return await Olaf.exportFileURL(entries: visible)
    }

    /// Exports the visible entries as **raw NDJSON** (for jq/backend analysis).
    public func exportNDJSONFileURL() async -> URL? {
        let visible = Array(filteredEntries.reversed())
        return await Olaf.exportNDJSONFileURL(entries: visible)
    }

    /// Writes the visible **network** entries to a HAR 1.2 file (opens in Charles/Proxyman/DevTools).
    public func exportHARFileURL() async -> URL? {
        let visible = Array(filteredEntries.reversed())
        return await Task.detached(priority: .utility) {
            guard let text = HARExporter.harDocument(from: visible) else { return nil }
            return LogExportFile.write(text, fileExtension: "har")
        }.value
    }

    /// Writes the visible **network** entries to a Postman Collection v2.1 file
    /// (same method+URL once; can be re-run via Postman → Import).
    public func exportPostmanFileURL() async -> URL? {
        let visible = Array(filteredEntries.reversed())
        return await Task.detached(priority: .utility) {
            guard let text = PostmanExporter.collection(from: visible) else { return nil }
            return LogExportFile.write(text, fileExtension: "postman_collection.json")
        }.value
    }

    /// Registered external tools (switch-to buttons).
    public var externalTools: [any ExternalToolBridge] {
        ExternalToolRegistry.shared.all
    }
}

File: Android/olaf/src/main/java/com/olaf/ui/model/LogViewerModel.kt
package com.olaf.ui.model

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.olaf.LogCategory
import com.olaf.LogEntry
import com.olaf.LogLevel
import com.olaf.Olaf
import com.olaf.internal.LogExportFile
import com.olaf.ui.util.HarExporter
import com.olaf.ui.util.PostmanExporter
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.FlowPreview
import kotlinx.coroutines.Job
import kotlinx.coroutines.withContext
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.debounce
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.io.File
import java.time.Instant

/**
 * State owner for the viewer: merges the instantaneous snapshot with the live stream and applies
 * level/category/content-type/text filtering.
 */
@OptIn(FlowPreview::class)
internal class LogViewerModel : ViewModel() {

    /** Display scope. */
    enum class Scope {
        /** This session — the in-memory ring buffer. */
        SESSION,

        /** The whole on-disk history, including previous sessions. */
        HISTORY
    }

    /** Grouped logs for one app session. */
    data class LogSession(
        val id: String,
        val startedAt: Instant,
        /** Newest first. */
        val entries: List<LogEntry>
    )

    // MARK: - Inputs

    private val _entries = MutableStateFlow<List<LogEntry>>(emptyList())

    private val _scope = MutableStateFlow(Scope.SESSION)
    val scope: StateFlow<Scope> = _scope

    private val _searchText = MutableStateFlow("")
    val searchText: StateFlow<String> = _searchText

    private val _enabledLevels = MutableStateFlow(LogLevel.entries.toSet())
    val enabledLevels: StateFlow<Set<LogLevel>> = _enabledLevels

    private val _selectedCategories = MutableStateFlow(emptySet<LogCategory>())
    val selectedCategories: StateFlow<Set<LogCategory>> = _selectedCategories

    private val _selectedContentKinds = MutableStateFlow(emptySet<NetworkContentKind>())
    val selectedContentKinds: StateFlow<Set<NetworkContentKind>> = _selectedContentKinds

    /** While `true` new logs keep flowing in; `false` freezes the list. */
    private val _isFollowing = MutableStateFlow(true)
    val isFollowing: StateFlow<Boolean> = _isFollowing

    /** Pinned entries, shown in their own section above the list, independent of filters. */
    private val _pinnedIds = MutableStateFlow(emptySet<String>())
    val pinnedIds: StateFlow<Set<String>> = _pinnedIds

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading

    private val _isLoadingMore = MutableStateFlow(false)
    val isLoadingMore: StateFlow<Boolean> = _isLoadingMore

    private val _hasMoreHistory = MutableStateFlow(false)
    val hasMoreHistory: StateFlow<Boolean> = _hasMoreHistory

    // MARK: - Internal state

    private var pendingWhilePaused = mutableListOf<LogEntry>()
    private var streamJob: Job? = null
    private var historyCursor: String? = null
    private var appliedDefaultCategory = false

    // MARK: - Derived state

    /** Debounced, normalized search query, so a keystroke doesn't rescan the whole list. */
    private val effectiveQuery: StateFlow<String> = _searchText
        .map { it.trim().lowercase() }
        .distinctUntilChanged()
        .debounce(SEARCH_DEBOUNCE_MS)
        .stateIn(viewModelScope, SharingStarted.Eagerly, "")

    /**
     * Filtered entries plus the decode index, derived in one pass so the two can never disagree —
     * a row hidden by one but not badged by the other would make a decode failure unfindable.
     */
    private val derived: StateFlow<Pair<List<LogEntry>, DecodeAttachmentIndex>> = combine(
        _entries,
        effectiveQuery,
        _enabledLevels,
        _selectedCategories,
        _selectedContentKinds
    ) { entries, query, levels, categories, contentKinds ->
        val index = DecodeAttachmentIndex.build(entries)
        filter(entries, query, levels, categories, contentKinds, index) to index
    }.stateIn(viewModelScope, SharingStarted.Eagerly, emptyList<LogEntry>() to DecodeAttachmentIndex.Empty)

    /** Filtered entries, newest first. */
    val filteredEntries: StateFlow<List<LogEntry>> = derived
        .map { it.first }
        .stateIn(viewModelScope, SharingStarted.Eagerly, emptyList())

    /** Decode errors folded under their network entry. */
    val decodeIndex: StateFlow<DecodeAttachmentIndex> = derived
        .map { it.second }
        .stateIn(viewModelScope, SharingStarted.Eagerly, DecodeAttachmentIndex.Empty)

    /** Categories present in the loaded entries — drives the chip bar. */
    val availableCategories: StateFlow<List<LogCategory>> = _entries
        .map { categoriesIn(it) }
        .distinctUntilChanged()
        .stateIn(viewModelScope, SharingStarted.Eagerly, emptyList())

    /** Pinned entries, newest first, independent of the filters. */
    val pinnedEntries: StateFlow<List<LogEntry>> = combine(_entries, _pinnedIds) { entries, ids ->
        if (ids.isEmpty()) emptyList() else entries.asReversed().filter { it.id in ids }
    }.stateIn(viewModelScope, SharingStarted.Eagerly, emptyList())

    /** History sessions, excluding the current one — the History tab groups by these. */
    val sessionGroups: StateFlow<List<LogSession>> = filteredEntries
        .map { groupSessions(it, Olaf.currentSessionId) }
        .stateIn(viewModelScope, SharingStarted.Eagerly, emptyList())

    /** Is a non-default filter applied? Drives the funnel badge. */
    val isFiltering: StateFlow<Boolean> = combine(
        _enabledLevels,
        _selectedCategories,
        _selectedContentKinds
    ) { levels, categories, contentKinds ->
        levels.size != LogLevel.entries.size || categories.isNotEmpty() || contentKinds.isNotEmpty()
    }.stateIn(viewModelScope, SharingStarted.Eagerly, false)

    // MARK: - Lifecycle

    fun start() {
        reload()
        if (streamJob != null) return
        streamJob = viewModelScope.launch {
            Olaf.stream().collect(::append)
        }
    }

    /**
     * (Re)loads entries for the current scope. Both reads are asynchronous, so the UI thread is
     * never blocked — even while the store is working through a burst. In history mode only the
     * FIRST page is loaded; older pages arrive through [loadOlderHistory].
     */
    fun reload() {
        pendingWhilePaused.clear()
        historyCursor = null
        _hasMoreHistory.value = false
        _isLoadingMore.value = false

        viewModelScope.launch {
            when (_scope.value) {
                Scope.SESSION -> {
                    _entries.value = Olaf.snapshotAsync()
                }

                Scope.HISTORY -> {
                    _isLoading.value = true
                    val page = Olaf.loadPersistedPage(minimumEntries = HISTORY_PAGE_SIZE)
                    _entries.value = page.entries
                    historyCursor = page.nextCursor
                    _hasMoreHistory.value = page.nextCursor != null
                    _isLoading.value = false
                }
            }
            applyDefaultCategorySelection()
        }
    }

    /** Loads the next (older) page of history. A no-op while loading or once history runs out. */
    fun loadOlderHistory() {
        val cursor = historyCursor ?: return
        if (_scope.value != Scope.HISTORY || _isLoading.value || _isLoadingMore.value) return

        _isLoadingMore.value = true
        viewModelScope.launch {
            val page = Olaf.loadPersistedPage(cursor, minimumEntries = HISTORY_PAGE_SIZE)
            // Drop the result if a reload (scope change, clear) landed in the meantime.
            if (_scope.value != Scope.HISTORY || historyCursor != cursor) {
                _isLoadingMore.value = false
                return@launch
            }
            // `_entries` runs oldest to newest, so an older page goes to the FRONT.
            _entries.update { page.entries + it }
            historyCursor = page.nextCursor
            _hasMoreHistory.value = page.nextCursor != null
            _isLoadingMore.value = false
        }
    }

    fun setScope(newScope: Scope) {
        if (newScope == _scope.value) return
        _scope.value = newScope
        reload()
    }

    private fun append(entry: LogEntry) {
        // Live entries aren't mixed into history scope — the current session isn't shown there
        // anyway, and switching back to Session reloads the latest snapshot.
        if (_scope.value == Scope.HISTORY) return

        if (_isFollowing.value) {
            _entries.update { it + entry }
            applyDefaultCategorySelection()
        } else {
            pendingWhilePaused.add(entry)
        }
    }

    fun pauseFollowing() {
        _isFollowing.value = false
    }

    fun resumeFollowing() {
        _isFollowing.value = true
        if (pendingWhilePaused.isNotEmpty()) {
            val flushed = pendingWhilePaused.toList()
            pendingWhilePaused.clear()
            _entries.update { it + flushed }
        }
    }

    /**
     * The viewer opens with the `network` chip preselected — but only once network entries
     * actually exist, because the chip bar only renders categories that were seen: preselecting an
     * unseen one would blank the list behind an invisible filter. Never overrides a user choice.
     */
    private fun applyDefaultCategorySelection() {
        if (appliedDefaultCategory || _selectedCategories.value.isNotEmpty()) return
        val entries = _entries.value
        if (entries.isEmpty()) return
        appliedDefaultCategory = true
        _selectedCategories.value = defaultCategorySelection(entries)
    }

    // MARK: - Actions

    fun setSearchText(value: String) {
        _searchText.value = value
    }

    fun toggleCategory(category: LogCategory) {
        _selectedCategories.update { if (category in it) it - category else it + category }
    }

    fun toggleLevel(level: LogLevel) {
        _enabledLevels.update { if (level in it) it - level else it + level }
    }

    fun toggleContentKind(kind: NetworkContentKind) {
        _selectedContentKinds.update { if (kind in it) it - kind else it + kind }
    }

    fun togglePin(entry: LogEntry) {
        _pinnedIds.update { if (entry.id in it) it - entry.id else it + entry.id }
    }

    fun resetFilters() {
        _enabledLevels.value = LogLevel.entries.toSet()
        _selectedCategories.value = emptySet()
        _selectedContentKinds.value = emptySet()
    }

    fun clear() {
        Olaf.clear()
        _entries.value = emptyList()
        pendingWhilePaused.clear()
        _pinnedIds.value = emptySet()
    }

    // MARK: - Export

    /**
     * Writes the entries **currently on screen** (scope plus every filter) to a shareable file,
     * chronologically — [filteredEntries] is newest first, so it is reversed.
     */
    suspend fun exportLogFile(): File? = Olaf.exportFile(filteredEntries.value.asReversed())

    /** Exports the visible entries as raw NDJSON, for `jq` and other tooling. */
    suspend fun exportNdjsonFile(): File? = Olaf.exportNdjsonFile(filteredEntries.value.asReversed())

    /** Writes the selected entries (multi-select) to a readable `.log` file, chronologically. */
    suspend fun exportSelected(ids: Set<String>): File? {
        val chosen = filteredEntries.value.filter { it.id in ids }.asReversed()
        return if (chosen.isEmpty()) null else Olaf.exportFile(chosen)
    }

    /** Writes the visible **network** entries as HAR 1.2 — opens in Charles/Proxyman/DevTools. */
    suspend fun exportHarFile(cacheDirectory: File): File? = withContext(Dispatchers.IO) {
        val text = HarExporter.harDocument(filteredEntries.value.asReversed())
        LogExportFile.write(cacheDirectory, text, fileExtension = "har")
    }

    /** Writes the visible **network** entries as a Postman Collection v2.1. */
    suspend fun exportPostmanFile(cacheDirectory: File): File? = withContext(Dispatchers.IO) {
        val text = PostmanExporter.collection(filteredEntries.value.asReversed())
        LogExportFile.write(cacheDirectory, text, fileExtension = "postman_collection.json")
    }

    /** Statistics over the entries currently on screen. */
    fun statistics(): NetworkStats = NetworkStats.compute(filteredEntries.value)

    // MARK: - Derivation logic (pure functions → directly testable)

    internal companion object {
        private const val SEARCH_DEBOUNCE_MS = 200L
        private const val HISTORY_PAGE_SIZE = 500

        /**
         * Level/category/content-type/search filter — newest first.
         *
         * An attached decode entry is never listed on its own; its network row answers for it
         * instead, matching the `decoding` chip, the `ERROR` level and any query that hits the
         * attached entry — so folding never makes a decode failure unfindable.
         */
        fun filter(
            entries: List<LogEntry>,
            query: String, // already trimmed and lowercased by the debounce pipeline
            levels: Set<LogLevel>,
            categories: Set<LogCategory>,
            contentKinds: Set<NetworkContentKind> = emptySet(),
            decodeIndex: DecodeAttachmentIndex = DecodeAttachmentIndex.Empty
        ): List<LogEntry> = entries.asReversed().filter { entry ->
            if (entry.id in decodeIndex.attachedIds) return@filter false
            val attached = decodeIndex.errors(entry)

            if (levels.isNotEmpty() && entry.level !in levels) {
                if (attached.isEmpty() || LogLevel.ERROR !in levels) return@filter false
            }
            if (categories.isNotEmpty() && entry.category !in categories) {
                if (attached.isEmpty() || LogCategory.Decoding !in categories) return@filter false
            }
            if (contentKinds.isNotEmpty()) {
                val kind = NetworkContentKind.of(entry)
                if (kind == null || kind !in contentKinds) return@filter false
            }
            query.isEmpty() || matchesQuery(entry, query) || attached.any { matchesQuery(it, query) }
        }

        fun matchesQuery(entry: LogEntry, query: String): Boolean {
            if (entry.message.lowercase().contains(query)) return true
            if (entry.category.rawValue.lowercase().contains(query)) return true
            return entry.metadata.any { (key, value) ->
                key.lowercase().contains(query) || value.lowercase().contains(query)
            }
        }

        /**
         * Groups entries by session, skipping [current]. Sessions come back newest first, and so
         * do the entries within each one.
         */
        fun groupSessions(entries: List<LogEntry>, current: String): List<LogSession> {
            val grouped = LinkedHashMap<String, MutableList<LogEntry>>()
            for (entry in entries) {
                if (entry.sessionId == current) continue
                grouped.getOrPut(entry.sessionId) { mutableListOf() }.add(entry)
            }
            return grouped.map { (id, items) ->
                LogSession(
                    id = id,
                    startedAt = items.minOf { it.date },
                    entries = items
                )
            }.sortedByDescending { it.startedAt }
        }

        /** `network` is preselected when network entries exist, nothing otherwise. */
        fun defaultCategorySelection(entries: List<LogEntry>): Set<LogCategory> =
            if (entries.any { it.category == LogCategory.Network }) setOf(LogCategory.Network) else emptySet()

        /** Categories seen in the entries, deduplicated and sorted by name. */
        fun categoriesIn(entries: List<LogEntry>): List<LogCategory> =
            entries.mapTo(LinkedHashSet()) { it.category }.sortedBy { it.rawValue }
    }
}
